std/format.d

```diff
@@ -1735,10 +1735,22 @@ FormatSpec!Char singleSpec(Char)(Char[] fmt)
  * `toString` should have one of the following signatures:
  *
  * ---
- * const void toString(scope void delegate(const(char)[]) sink, FormatSpec fmt);
- * const void toString(scope void delegate(const(char)[]) sink, string fmt);
- * const void toString(scope void delegate(const(char)[]) sink);
- * const string toString();
+ * void toString(W)(W w, FormatSpec fmt)
+ * void toString(W)(W w)
+ * string toString();
+ * ---
+ *
+ * Where `W` allows a character accepting
+ * $(REF_ALTTEXT ouput range, isOutputRange, std,_range,primitives) with its template constraints.
+ * The template type does not have to be called `W`.
+ *
+ * The following overloads are also accepted for legacy reasons. It's recommended that
+ * any new code forgo these overloads for speed and attribute acceptance reasons.
```

Comment:
What about also accepting a `toString` overload that takes an output range that can accept any range of characters?  If efficiency is a concern here, it would make sense for `toString` to be able to take a, output range of char/wchar/dchar instead of needing a function call roundtrip per character.

Reply:
Is it really just for legacy reasons though? What about classes? As soon as you templatize `toString`, it can't be virtual.

Reply:
Hmm, that's true. Virtual `toString` methods could be problematic because `format` may not know how to construct an output range of a type that will be accepted by `toString`.

Perhaps one way out is use `std.range.interfaces.OutputRangeObject`. Since such an object is an output range, it can be recognized by `format`, and since it gives an opaque interface to arbitrary underlying output range implementations, it will also give us the flexibility to choose the best implementation used by `format`.

OTOH, since `scope void delegate(const(char)[])` is already established in current code, and already allows arbitrary implementations, perhaps we should just retain support for it.  Code that can afford to have `toString` templatized can still take advantage of a more flexible implementation of `format`.

Reply:
`@safe`ty is still an issue for virtual `toString` functions, though. Well, as well as other attributes.  I've no idea how to work around that.